File: neetCode150/py/236LowestCommonAncestorofaBinaryTree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        list1 = []
        list2 = []
        self.dfs(root,p,list1)
        self.dfs(root,q,list2)
        for i in list1:
            for j in list2:
                if i == j:
                    return i
    
    def dfs(self, root, target, list):
        if not root:
            return False
        if root == target:
            list.insert(0,root)
            return True
        list.insert(0,root)
        if self.dfs(root.left, target, list):
            return True
        else:
            list.pop(0)
        list.insert(0,root)
        if self.dfs(root.right, target, list):
            return True
        else:
            list.pop(0)
        return False

# Create a binary tree
root = TreeNode(3)
root.left = TreeNode(5)
root.right = TreeNode(1)
root.left.left = TreeNode(6)
root.left.right = TreeNode(2)
root.right.left = TreeNode(0)
root.right.right = TreeNode(8)
root.left.right.left = TreeNode(7)
root.left.right.right = TreeNode(4)

# Initialize the solution
solution = Solution()

# Find the lowest common ancestor of nodes with values 5 and 1
ancestor = solution.lowestCommonAncestor(root, root.left, root.right)
print("Lowest Common Ancestor:", ancestor.val)

File: 2025/Q4LeetCode/236_LowestCommonAncestorofaBinaryTree.js
// 236. Lowest Common Ancestor of a Binary Tree

class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = this.right = null;
  }
}

const lowestCommonAncestor = (root, p, q) => {
  if (!root || root === p || root === q) return root;

  const left = lowestCommonAncestor(root.left, p, q);
  const right = lowestCommonAncestor(root.right, p, q);

  if (left && right) return root;
  return left || right;
};

// Example usage:
const root = new TreeNode(3);
root.left = new TreeNode(5);
root.right = new TreeNode(1);
root.left.left = new TreeNode(6);
root.left.right = new TreeNode(2);
root.right.left = new TreeNode(0);
root.right.right = new TreeNode(8);
root.left.right.left = new TreeNode(7);
root.left.right.right = new TreeNode(4);

const p = root.left; // Node with value 5
const q = root.right; // Node with value 1

const lca = lowestCommonAncestor(root, p, q);
console.log(lca.val); // Output: 3
